### PR TITLE
log on action and prob for off-policy evaluation

### PR DESCRIPTION
**Pull request by jonastim**
The main change is for the off-policy evaluator to log the action and probability of the learning model rather than that of the logged data (which is identical for all models and not very useful when trying to compare different learners).

Introduced a helper function for sampling the actions and used if in some other code placed to avoid redundancy.

### CODE REVIEW

File: coba/evaluators/online.py [hunk 1]
```diff
@@ -235,8 +235,10 @@ def evaluate(self, environment: Optional[Environment], learner: Optional[Learner
                         predict_time = time.time()-start_time
                         if not batched:
                             ope_reward = sum(p*float(log_rewards.eval(a)) for p,a in zip(on_probs,log_actions))
+                            on_action, on_prob = sample_actions(log_actions, on_probs)
                         else:
                             ope_reward = [ sum(p*float(R.eval(a)) for p,a in zip(P,A)) for P,A,R in zip(on_probs,log_actions,log_rewards) ]
+                            on_action, on_prob = zip(*[sample_actions(actions, probs) for actions, probs in zip(log_actions, on_probs)])
                     else:
```
Comment (jonastim) [inline coba/evaluators/online.py:242]: How to do this for continuous actions?

File: coba/learners/safety.py [hunk 2]
```diff
@@ -287,9 +287,7 @@ def predict(self, context: Context, actions: Actions) -> Tuple[Action,Prob,kwarg
 
             A,P = [],[]
             if self._pred_format[:2] == 'PM':
-                I = [self._get_pmf_index(p) for p in pred]
-                A = [ a[i] for a,i in zip(actions,I) ]
-                P = [ p[i] for p,i in zip(pred,I) ]
+                A, P = list(map(list, zip(*[sample_actions(a, p, self._rng) for a, p in zip(actions, pred)])))
```
Comment (jonastim) [inline coba/learners/safety.py:290]: Could remove `list(map(list,` if it was ok to return a tuple instead of a list.

**Comment by jonastim** [comment]
I can't seem to re-run tests. 
Looks like these unrelated tests should be less sensitive
```
======================================================================
FAIL: test_DM (coba.tests.test_environments_filters.OpeRewards_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/coba/coba/coba/tests/test_environments_filters.py", line 2390, in test_DM
    self.assertAlmostEqual(new_interactions[0]['rewards'].eval('c'),.79699, places=4)
AssertionError: 0.7970473766326904 != 0.79699 within 4 places (5.7376632690453455e-05 difference)

======================================================================
FAIL: test_DM_action_not_hashable (coba.tests.test_environments_filters.OpeRewards_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/coba/coba/coba/tests/test_environments_filters.py", line 2404, in test_DM_action_not_hashable
    self.assertAlmostEqual(new_interactions[0]['rewards'].eval(['c']),.79699, places=4)
AssertionError: 0.7970473766326904 != 0.79699 within 4 places (5.7376632690453455e-05 difference)

----------------------------------------------------------------------
```

File: coba/evaluators/online.py (hunk 1 above)
Comment (mrucker) [reply coba/evaluators/online.py:242]: For continuous actions we just need to call `on_action,on_prob = predict(log_context, log_actions)[:2]`

Comment (mrucker) [reply coba/evaluators/online.py:242]: Maybe on line 246? I don't think we need to have separate processing for batched and non-batched. Man I hate all this batched logic. It's all here for neural network stuff we do where backpropagation with mini-batches can give huge gains in computation time.

**Comment by mrucker** [comment]
> I can't seem to re-run tests. Looks like these unrelated tests should be less sensitive
> 
> ```
> ======================================================================
> FAIL: test_DM (coba.tests.test_environments_filters.OpeRewards_Tests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/runner/work/coba/coba/coba/tests/test_environments_filters.py", line 2390, in test_DM
>     self.assertAlmostEqual(new_interactions[0]['rewards'].eval('c'),.79699, places=4)
> AssertionError: 0.7970473766326904 != 0.79699 within 4 places (5.7376632690453455e-05 difference)
> 
> ======================================================================
> FAIL: test_DM_action_not_hashable (coba.tests.test_environments_filters.OpeRewards_Tests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/runner/work/coba/coba/coba/tests/test_environments_filters.py", line 2404, in test_DM_action_not_hashable
>     self.assertAlmostEqual(new_interactions[0]['rewards'].eval(['c']),.79699, places=4)
> AssertionError: 0.7970473766326904 != 0.79699 within 4 places (5.7376632690453455e-05 difference)
> 
> ----------------------------------------------------------------------
> ```

Totally agree. Are these the only tests you're having problems with? I've tried to bump down a lot of tests over time.

File: coba/tests/test_evaluators_online.py [hunk 3]
```diff
@@ -658,8 +658,17 @@ def request(self, context, actions, request):
     def test_batched_request_continuous(self):
         class TestLearner:
             def request(self,context,actions,request):
-                if isinstance(context,BatchType): raise Exception()
-                return .5
+                if isinstance(context,BatchType):
+                    raise Exception()
+                return 0.5
+
+            def predict(self, context, actions):
```
Comment (jonastim) [inline coba/tests/test_evaluators_online.py:665]: Struggling to make this test pass. 
The processing thinks it's of `AX` format and then always fills in `None`s for the probability. I haven't worked with continuous actions before and I am not quite sure about all the different formats and in the `SafeLearner`
Any advice, @mrucker?

Screenshot 2023-11-06 at 12 32 28 PM

File: coba/evaluators/online.py (hunk 1 above)
Comment (jonastim) [reply coba/evaluators/online.py:242]: Tried to add support for continuous actions but struggle to make some tests pass, see below.

**Pull request merged**